clients/python/test/e2e/ci/setup_e2e_pytest.py

```diff
@@ -0,0 +1,121 @@
+from urllib.parse import urlparse, ParseResult
+from typing import List, Dict, Any, Union
+import typer
+from packaging import version
+import json
+from _utils import E2eScriptFailure, E2eExitCodes
+import warnings
+
+# keys to be found in input dicts
+surl: str = "OSPARC_API_HOST"
+skey: str = "OSPARC_API_KEY"
+ssecret: str = "OSPARC_API_SECRET"
+skeys: List[str] = [surl, skey, ssecret]
+
+crepo: str = "OSPARC_CLIENT_REPO"
+cbranch: str = "OSPARC_CLIENT_BRANCH"
+cversion: str = "OSPARC_CLIENT_VERSION"
+ckeys: List[str] = [crepo, cbranch, cversion]
+
+
+def main(client_config: str, server_config: str) -> bool:
+    """
+    Generates a toml configuration file pytest e2e tests
+
+    exceptions:
+    -----------
+        A typer.Exit(code=100) is raised if a failure is encountered
+
+    returns:
+    --------
+        A bool indicating whether or not the (client, server) pair are compatible
+    """
+    # read in data
+    ccfg = json.loads(client_config)
+    scfg = json.loads(server_config)
+    if not isinstance(ccfg, dict):
+        warnings.warn(
+            f"The client configuration received in {__file__} was invalid",
+            E2eScriptFailure,
+        )
+        typer.Exit(code=E2eExitCodes.CI_SCRIPT_FAILURE)
+    if not isinstance(scfg, dict):
+        warnings.warn(
+            f"The server configuration received in {__file__} was invalid",
+            E2eScriptFailure,
+        )
+        typer.Exit(code=E2eExitCodes.CI_SCRIPT_FAILURE)
+    if not all(key in skeys for key in scfg.keys()):
+        warnings.warn(
+            f"The following server inputs are required: {skeys}. Received: {set(scfg.keys())}",
+            E2eScriptFailure,
+        )
+        typer.Exit(code=E2eExitCodes.CI_SCRIPT_FAILURE)
+
+    osparc_url: ParseResult = urlparse(scfg[surl])
+    ini_file: Path = Path(__file__).parent.parent / "pyproject.toml"
+    ini_file.unlink(missing_ok=True)
+    comp_df: pd.DataFrame = pd.read_json(
+        Path(__file__).parent.parent / "data" / "server_client_compatibility.json"
+    )
+
+    # sanity check client inputs
+    if not (ccfg[cbranch] == "" or ccfg[cversion] == ""):
+        warnings.warn(
+            f"{cbranch}={ccfg[cbranch]}, {cversion}={ccfg[cversion]}", E2eScriptFailure
+        )
+        raise typer.Exit(code=E2eExitCodes.CI_SCRIPT_FAILURE)
+    # sanity checks
+    client_ref: str
+    if ccfg[cversion] != "":
+        client_ref = "production"
+    else:
+        if ccfg[crepo] == "":
+            warnings.warn(
+                f"{cbranch}={ccfg[cbranch]}, {crepo}={ccfg[crepo]}", E2eScriptFailure
+            )
+            raise typer.Exit(code=E2eExitCodes.CI_SCRIPT_FAILURE)
+        else:
+            client_ref = ccfg[cbranch]
+
+    # set environment variables
+    envs: List[str] = []
+    envs.append(f"OSPARC_API_HOST = {scfg[surl]}")
+    envs.append(f"OSPARC_API_KEY = {scfg[skey]}")
+    envs.append(f"OSPARC_API_SECRET = {scfg[ssecret]}")
+
+    pytest_settings: Dict[str, Any] = {}
```

Comment:
TIP: from 3.8 you can use normal `dict` to annotate, i.e.
```python
pytest_settings: dict[str, Any] = {}
```

Reply:
fix